rest/data/host_create.go

```diff
@@ -0,0 +1,228 @@
+			return errors.Wrap(err, "unable to parse command")
+		}
+		for _, cmd := range cmds {
+			createHost, ok := cmd.(*command.CreateHost)
+			if !ok {
+				continue
+			}
+			if err = createHost.CreateHost.Validate(); err != nil {
+				catcher.Add(err)
+				continue
+			}
+			createHostCmds = append(createHostCmds, createHost)
+		}
+	}
+	if catcher.HasErrors() {
+		return catcher.Resolve()
+	}
+
+	hosts := []host.Host{}
+	for _, createHost := range createHostCmds {
+		for i := 0; i < createHost.CreateHost.NumHosts; i++ {
+			intent, err := dc.MakeIntentHost(t.Id, user.Username(), keyVal, *createHost.CreateHost)
+			if err != nil {
+				return errors.Wrap(err, "error creating host document")
+			}
+			hosts = append(hosts, *intent)
+		}
+	}
+
+	return errors.Wrap(host.InsertMany(hosts), "error inserting host documents")
+}
+
+func (dc *DBCreateHostConnector) MakeIntentHost(taskID, userID, publicKey string, createHost apimodels.CreateHost) (*host.Host, error) {
+	provider := evergreen.ProviderNameEc2OnDemand
+	if createHost.Spot {
+		provider = evergreen.ProviderNameEc2Spot
+	}
+
+	// get distro if it is set
+	d := distro.Distro{}
+	ec2Settings := cloud.EC2ProviderSettings{}
+	var err error
+	if distroID := createHost.Distro; distroID != "" {
+		d, err = distro.FindOne(distro.ById(distroID))
+		if err != nil {
+			return nil, errors.Wrap(err, "problem finding distro")
+		}
+		if err := mapstructure.Decode(d.ProviderSettings, &ec2Settings); err != nil {
+			return nil, errors.Wrap(err, "problem unmarshaling provider settings")
+		}
+	}
+
+	// set provider
+	d.Provider = provider
+
+	if publicKey != "" {
+		d.Setup += fmt.Sprintf("\necho \"\n%s\" >> ~%s/.ssh/authorized_keys\n", publicKey, d.User)
+	}
+
+	// set provider settings
+	if createHost.AMI != "" {
+		ec2Settings.AMI = createHost.AMI
+	}
+	if createHost.AWSKeyID != "" {
+		ec2Settings.AWSKeyID = createHost.AWSKeyID
+		ec2Settings.AWSSecret = createHost.AWSSecret
+	}
+
+	for _, mount := range createHost.EBSDevices {
+		ec2Settings.MountPoints = append(ec2Settings.MountPoints, cloud.MountPoint{
+			DeviceName: mount.DeviceName,
+			Size:       int64(mount.SizeGiB),
+			Iops:       int64(mount.IOPS),
+			SnapshotID: mount.SnapshotID,
+		})
+	}
+	if createHost.InstanceType != "" {
+		ec2Settings.InstanceType = createHost.InstanceType
+	}
+	if createHost.KeyName != "" {
+		ec2Settings.KeyName = createHost.KeyName
+	}
+	if createHost.Region != "" {
+		ec2Settings.Region = createHost.Region
+	}
+	if len(createHost.SecurityGroups) > 0 {
+		ec2Settings.SecurityGroupIDs = createHost.SecurityGroups
+	}
+	if createHost.Subnet != "" {
+		ec2Settings.SubnetId = createHost.Subnet
+	}
+	if createHost.UserdataCommand != "" {
+		ec2Settings.UserData = createHost.UserdataCommand
+	}
+	if createHost.VPC != "" {
+		ec2Settings.VpcName = createHost.VPC
+	}
+	if err := mapstructure.Decode(ec2Settings, &d.ProviderSettings); err != nil {
+		return nil, errors.Wrap(err, "error marshaling provider settings")
+	}
```

Comment:
I feel like I've read this validation elsewhere. Is there somewhere better we can validate thise. 

Furthermore, it seems likely that we're going to want to do this for other host types, so putting ec2 specific logic here seems like it's going to be rough in not very long.

Reply:
It's not clear from the diff, but I moved this function from the route package to the data package without changing it at all. It's not duplicated - what you remember is seeing it there